qiskit/optimization/algorithms/grover_minimum_finder.py

```diff
@@ -0,0 +1,309 @@
+        opt_prob_converter = OptimizationProblemToNegativeValueOracle(n_value,
+                                                                      backend=self._backend)
+
+        while not optimum_found:
+            m = 1
+            improvement_found = False
+
+            # Get oracle O and the state preparation operator A for the current threshold.
+            problem_.objective.set_offset(orig_constant - threshold)
+            a_operator, oracle, func_dict = opt_prob_converter.encode(problem_)
+
+            # Iterate until we measure a negative.
+            loops_with_no_improvement = 0
+            while not improvement_found:
+                # Determine the number of rotations.
+                loops_with_no_improvement += 1
+                rotation_count = int(np.ceil(random.uniform(0, m-1)))
+                rotations += rotation_count
+
+                # Apply Grover's Algorithm to find values below the threshold.
+                if rotation_count > 0:
+                    grover = Grover(oracle, init_state=a_operator, num_iterations=rotation_count)
+                    circuit = grover.construct_circuit(
+                        measurement=self._backend.name() != "statevector_simulator")
+                else:
+                    circuit = a_operator._circuit
+
+                # Get the next outcome.
+                outcome = self._measure(circuit, n_key, n_value, self._backend)
+                k = int(outcome[0:n_key], 2)
+                v = outcome[n_key:n_key + n_value]
+
+                # Convert the binary string to integer.
+                int_v = self._bin_to_int(v, n_value) + threshold
+                v = self._twos_complement(int_v, n_value)
+
+                self._logger.info('Iterations: {}'.format(rotation_count))
+                self._logger.info('Outcome: {}'.format(outcome))
+                self._logger.info('Value: {} = {}'.format(v, int_v))
+
+                # If the value is an improvement, we update the iteration parameters (e.g. oracle).
+                if int_v < optimum_value:
+                    optimum_key = k
+                    optimum_value = int_v
+                    self._logger.info('Current Optimum Key: {}'.format(optimum_key))
+                    self._logger.info('Current Optimum Value: {}'.format(optimum_value))
+                    if v.startswith("1"):
+                        improvement_found = True
+                        threshold = optimum_value
+                else:
+                    # No better number after the max number of iterations, so we assume the optimal.
+                    if loops_with_no_improvement >= self._n_iterations:
+                        improvement_found = True
+                        optimum_found = True
+
+                    # Using Durr and Hoyer method, increase m.
+                    # TODO: Give option for a rotation schedule, or for different lambda's.
+                    m = int(np.ceil(min(m * 8/7, 2**(n_key / 2))))
```

Comment:
Hi Austin, in this loop you are re-implementing Grover's algorithm with a different rotation schedule, right? How would we need to adapt `Grover` so that you could use it here? I think it's better if we use the Grover algorithm directly (probably also saves a few lines of code!).

Reply:
We aren't re-implementing Grover's algorithm, just calling it with different rotation counts at each iteration. The code in GroverMinimumFinder is based on the algorithmic framework Grover Adaptive Search.